akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -476,42 +495,45 @@ final class FoldAsync[In, Out](zero: Out, f: (Out, In) ⇒ Future[Out]) extends
               completeStage()
             } else if (isAvailable(out) && !hasBeenPulled(in)) tryPull(in)
           }
+      }.invoke _
+
+      def onPush(): Unit = {
+        try {
+          aggregating = f(aggregator, grab(in))
+          handleAggregatingValue()
+        } catch {
+          case NonFatal(ex) ⇒ decider(ex) match {
+            case Supervision.Stop ⇒ failStage(ex)
+            case supervision ⇒ {
+              supervision match {
+                case Supervision.Restart ⇒ onRestart(ex)
+                case _                   ⇒ () // just ignore on Resume
+              }
+
+              tryPull(in)
+            }
+          }
         }
       }
-    }).invoke _
 
-    def onPush(): Unit = {
-      try {
-        aggregating = f(aggregator, grab(in))
+      override def onUpstreamFinish(): Unit = {
+        handleAggregatingValue()
```

Comment:
@johanandren I have added this test: 
`
"complete future and return zero + item given a stream of one item" in assertAllStagesStopped {
      val futureValue =
        Source.single(100)
          .runFoldAsync(5)((acc, elem) ⇒ Future.successful(acc + elem))
      Await.result(futureValue, remainingOrDefault) should be(105)
    }
`
and it passes. Also I added 
`
        printf("Callback called\n")
`
at the beginning of the callback and the text is printed only once when running the test given above. Of course I'm not going to commit this printf, however it would be nice to be able to verify in the unit test the actual number of invocations of the callback, but I don't know how to do that on top of my head.